### ISSUE TITLE
Backends as plugins

### ISSUE THREAD

**Issue by marcinwyszynski** [opentffoundation/opentf#382]
### Summary

In this RFC I explore the idea of supporting external backends using a plugin system identical to what OpenTF currently supports for modules. Note that this is not a commitment that it would be done, or that it would be done in a way described in this RFC. The purpose of this document is to describe one potential approach, and gauge overall interest.

### Problem Statement

Since backends are an integral part of the OpenTF core, a user who wants or needs to upgrade to a new version of the backend (eg. supporting a new functionality they care about), they are forced to upgrade to a new version, which may be a significant amount of work for some cases.

There are also cases where a sophisticates user wants to use a system that's not currently supported, limiting their choices to forking the project, or building an HTTP server that implements the [http API](https://developer.hashicorp.com/terraform/language/settings/backends/http) and forwards the requests to the external system.

### User-facing description

The new feature could be built as a separate backend, thus not introducing any backwards incompatible changes.

An example usage could look like this:

```terraform
provider "plugin" {
  source  = "oci:ghcr.io/myaccount/myrepo"
  version = ">= .04.5"
  
  // plugin-specific config
}
```

### Technical Description

> Which components of OpenTF would be involved in the change?

Based on my preliminary research, we would need to need to create a new implementation of the [backend interface](https://github.com/opentffoundation/opentf/blob/main/internal/backend/backend.go#L58).

> How would the components need to change?

Ideally no existing components would **need to** change but it could be beneficial to add the backend checksum to the lockfile, similar to what we do for providers, for increased supply chain safety.

> Are there any interactions with other features that should be mentioned?

With this feature we're not bound by the existing registry API so we'd be free to forge our own path re: distribution. One option is to copy the provider API verbatim. The other option is to use this as an opportunity to test the [OCI approach](https://github.com/opentffoundation/opentf/issues/308).

### Rationale and alternatives

> What would be the impact of this feature?

Backends could be versioned and managed independently of OpenTF versions.
Users could provide backends specific to their needs.

> How would the OpenTF user experience suffer if we didn't do this.

It's just inconvenience, folks managed without it for years and while it sounds like a good idea overall, the existing status quo was good enough for the majority of users. 

### Downsides

Implementing this increases the complexity of the product and introduces new potential failure modes (eg. crashed backend).

**Comment by alrs** [comment]
Also: if backends are plugins, there's a chance that some popular backend could get switched to a non-free license, putting us right back where we started.

**Comment by marcinwyszynski** [comment]
> Unless the module interface is absolutely-unchanging

The state backend is in reality a super simple interface and sees very little change. 

> and forever-supported at earch version level,

I think it's a matter of backwards compatibility with supporting multiple protocol versions, not unlike what's currently happening for modules. Some newer versions of modules may not work with older versions of Terraform, but older modules should work with newer versions of Terraform.

> backend implementations will wind up pinned to arbitrarily new or old versions of OpenTF.

Again, I feel this is a matter of versioning protocols and supporting older versions for long time. Given how rarely the state API changes, I feel like that new protocol versions will be few and far between.

**Comment by marcinwyszynski** [comment]
> there's a chance that some popular backend could get switched to a non-free license, putting us right back where we started

Well, fair point, but this applies to any software. It doesn't feel like there would be major commercial interest in relicensing this kind of project, but worst case folks can fork the last true FOSS revision?

**Comment by clementd64** [comment]
> there's a chance that some popular backend could get switched to a non-free license

Providers and here backends usually work with external services. You will always be bounded to that services. For example the legacy remote backend use the closed source Terraform Cloud service. If the backend plugin change to a non free license it will change nothing about the closed source aspect of the main part of it.

> Well, fair point, but this applies to any software (including current providers).

Yes, unless owned by a non-profit organization with strict license policy, this risk exist everywhere.

**Comment by Hefeweizen** [comment]
@clementd64 asserts:
> You will always be bounded to that services.
"Always" seems a bit strong; "likely" sounds better.
I could imagine a git backend that could take any service; and while github and gitlab would be commercial, a self-host gitea instance would not.  Similarly, a backend for Ceph would not be to a single closed-source provider.

From that, I don't think the concern about closed-source backend implementations is unwarranted.

**Comment by Hefeweizen** [comment]
@marcinwyszynski in your mock, you have `provider` initializing the backend configuration.  I've not dug into the code, but in reading through the [architecture doc](https://github.com/opentffoundation/opentf/blob/main/docs/architecture.md) providers aren't mentioned until 'Graph Builder' which is after backend is called.  If it happens as part of Configuration Loader, then the diagram indicates provider-loading should happen independently.  Perhaps a dedicated keyword should be used?

But I'm probably jumping too far into this one approach when your overall goal was to gauge interest.

In that regard, do backends change so frequently that we need to decouple updates?  Perhaps that isn't a fair question, as absent the feature, nobody is doing it on a regular basis.  : )

If our reference to our backend repository is down, what happens to our application run?  Do we lose the ability to do anything?  I would hope a very conservative default is available so that end-users can continue to manage their infrastructure.

So, at this point, I'm intrigued, but in balancing reward vs risk, I'm not seeing the advantages this approach would provide in day-to-day usage.  From my perspective, I set backend once at the start of the project and then never really think about it again.  I want it to be stable and accessible.

**Comment by marcinwyszynski** [comment]
> you have provider initializing the backend configuration

Ah, Freudian slip, I was just thinking "it's going to look like a provider".
Fixed that, the block should obviously say "backend".

> In that regard, do backends change so frequently that we need to decouple updates?

As a matter of fact, they do sometimes change. Some are super stable like HTTP, but some like AWS see a decent bit of change, supporting an ever increasing range of options (eg. OIDC/web identity).

> If our reference to our backend repository is down, what happens to our application run? Do we lose the ability to do anything?

Fair point. But similar to how you can store provider plugins locally, you could store backend plugins, so if you can get the backend binary onto your system one way or another, you should be good I guess? Certainly something to keep in mind though.

**Comment by marcinwyszynski** [comment]
FYI I updated the RFC with a PoC of basic implementation.

**Comment by DCCoder90** [comment]
As long as there are no issues with backwards compatibility (which you've already addressed) I like this idea.  I feel like this would make everything much more flexible.  I can throw my hat into the ring in regard to general interest, but I believe there needs to be a wider discussion on this.

**Comment by Magnitus-** [comment]
As long as backwards api compatibility is maintained, I think its a wonderful idea.

While backends don't need to scale to the same extent as providers, I still think having backends coupled with the core is not ideal.

We having been using etcd as our backend for some time and given that project's history of supporting it in the core, that was a wild ride. First of all, we find out that states bigger than ~1.5mb are unsupported (because the state is stored in a single key). Then, the etcd backend gets deprecated in version 1.3.0.

So then, we roll out this bad boy which solves both problems: https://github.com/Ferlab-Ste-Justine/terraform-backend-etcd

But then, we need to change our backend paradigm from directly managed by opentf like providers to having to run an http proxy everywhere. And while we're doing that transition (which takes some time because we have a quazillion terraform pipelines and a lot of other stuff we need to do), we're stuck on opentf version 1.2.9 (last version that supported etcd in the core) and are not able to take advantage of cool new features like default field values in object variables.

The above proposal would help prevent insane situations like that from occuring.

**Comment by ggiallo28** [comment]
I appreciate the concept of interchangeable backends. Implementing the new solution doesn't exclude the possibility of retaining the existing backend, offering organizations the flexibility to tailor to their specific needs. I am eager to contribute to the development of this feature.

**Comment by huix11** [comment]
good

**Comment by cube2222** [comment]
I wonder if this should even be a special backend type.

If we look at providers, there is no
```
provider "plugin" {
  // ...
}
```
even though there are built-in providers and plugin providers.

My thinking would be to mirror the provider syntax.
```
backend "s3" {
  // ...
}
```
would just default to the `registry.opentofu.org/opentofu/s3` backend plugin.

In practice, if we do this, I'd totally be in favour of moving those cloud-specific backends (s3, gcs, azure, etc.) to their own plugins. I don't see any reason not to, as long as we make it fully backwards-compatible with no code changes required. They'd still be officially maintained by the OpenTofu team, of course.

In practice, we might want a block similar to `required_providers` that would be called `required_backend` where we can specify (optionally!) where to get the backend from. Something akin to
```
terraform {
  required_backend {
    source = "namespace/name"
  }
}
```

I think an approach like this would be much more consistent, syntax-wise.

**Comment by marcinwyszynski** [comment]
There is a subtle difference between backends and providers. Each module can list its required providers while only the root can specify a backend. Hence while I'm not married to any particular implementation or syntax, I think we should be drawing some distinction here.

**Comment by cube2222** [comment]
Actually, it's not just the root. The remote state datasource can be used to fetch state from various backends. So you could actually have multiple backend plugins active at a single time.

Generally, I'd want to use the exact same code for managing provider and backend plugins. Lockfiles, fetching from registry, etc. Factor it all out. It would be different URLs in the registry (like modules and providers) but they are the same concept technically.

Then, if we want, we can introduce alternative plugin hosting (like OCI registries), and have it work for all plugin types out of the box.

**Comment by marcinwyszynski** [comment]
> The remote state datasource can be used to fetch state from various backends.

Oh, right. Fair point!

**Comment by sftim** [comment]
:thought_balloon: could we implement backends as _internal_ plugins, initially? You would need to recompile to make a change but within the `tofu` binary, we start off with two communicating bits of code, one acting as the main OpenTofu and the other running the compiled-in code equivalent to the selected plugin?

That will let us shake out the idea and some of the problems, without leaping into supportability challenges (end users don't like having features dropped, even alpha features).

Go is quite good for implementing this kind of design, which is convenient.

**Comment by apparentlymart** [comment]
I see this touched on already in the discussion above, so I'm sorry for the slight redundancy, but I think it's worth noting that although the state storage API (which is represented internally by the [`statemgr.Full`](https://pkg.go.dev/github.com/opentofu/opentofu/internal/states/statemgr#Full) interface) hasn't changed for a long time, it has some annoying assumptions baked into it which I think we should be careful about committing to as a compatibility constraint for all time.

The most notable thing is the assumption that a state snapshot is _always_ (from the backend's perspective) a big opaque `[]byte` built by serializing a data structure that's shared with OpenTofu Core. This has a number of drawbacks, including but not limited to:

- Various casual use of deep copies to let the state managers safely run concurrently with ongoing work in OpenTofu Core even though they are _conceptually_ sharing the live state data structure. (https://github.com/opentofu/opentofu/issues/1580)
- Scales poorly as the total size of the state grows, because OpenTofu must always read and write the single huge blob rather than potentially making finer-grain updates if a particular backend were smart enough to do that. (For example, the `pg` backend _could_ use a separate row for each individual object tracked in the state, rather than one huge row for the entire state, if the API were designed differently.)
- Related to the previous point, it also creates some conflicting pressures to minimize how often we write (potentially large) interim state snapshots to the backend while also minimizing the amount of changes lost if something catastrophic happens during the apply phase, such as OpenTofu being killed by the Linux OOM killer.

With the API between core and the state storage backends currently entirely internal, we can in theory improve on any of these things at any time, although unfortunately that also means it tends to never be "the most important thing" and often gets deferred in favor of other work!

However, if we were to fix the current design in a documented protocol that external codebases depend on then we're likely to get stuck supporting it forever, even if we later introduce a new protocol that improves on the above design problems.

I don't have a concrete new design to propose here today, but the main broad idea I have in my mind is to let OpenTofu Core stream individual updates out to the state storage plugin and then let the plugin be the one to decide either to save the entire thing as a single blob like today, or to store many separate objects that can each be separately updated. The state storage plugin could then take full ownership of the decision about how often to commit changes to persistent storage, with _some_ plugins (like `pg` backend as previously mentioned) being able to perform immediate partial updates of their data store so that a catastrophic failure would lose as little data as physically possible.

While I certainly don't intend this comment to mean that we should _definitely_ sign up for a huge API redesign as a prerequisite for plugin state storage -- that wouldn't be very pragmatic -- I _do_ think we should weigh the potential benefits of implementing plugin-based state storage (vs. continuing to maintain all the backends in-tree for now) against the likelihood of wanting to change the design, to see if we'd prefer to prioritize at least a _partial_ rework of the state storage API before we fossilize it in a committed external API surface. If we can live with in-tree backend implementations for now then that can buy us more time to consider this question and decide whether to just "plugin-ize" the model we already have, or to try to adopt an improved API model first while we still have the luxury of updating all of the backends to the new API in one pass.

**Comment by apparentlymart** [comment]
I do broadly agree with you, and my comment was _at most_ intended to suggest a little delay so we can make this tradeoff rather than just _assuming_ that the current state storage API is the right one to commit to for the long term.

A potential middle-ground in the meantime is a model where OpenTofu will accept more backends into the codebase _for now_ on the condition that it's the vendor that acts as the primary maintainer of the backend, rather than the OpenTofu core team. That would admittedly force some changes to project governance to better support the situation where the project is co-maintained by several different entities, which is certainly simpler if you can simply assert that the parts with different maintainership are implemented in entirely separate repositories and released independently of the core. And I would agree that if we were adopting a more complex contribution model _only_ to support backends temporarily remaining in this repository then that would seem like throwaway work in the long run.

I don't wish to turn this into a big meta-discussion about project governance. That's a much broader question than this one, better discussed in a different place. :grinning: I'm just trying to put some different options on the table so we can figure out together what seems like the best tradeoff between short-term and long-term needs/goals.

(Naturally, I cannot comment on what might have motivated the current structure of things prior to the OpenTofu fork; I'm going to stay focused only on our options for how to move forward from where we are.)

**Comment by apparentlymart** [comment]
One thing we need to consider here that I don't think has been mentioned yet (sorry if I missed it):

We currently have a very intentional separation of concerns where `tofu init` only _installs_ plugins and never executes them, while other commands only _execute_ plugins and never install them. We have things arranged that way to allow for the following workflow:

1. Write or modify configuration to use a new plugin that wasn't previously used.
2. Run `tofu init`, at which point OpenTofu will install the plugin for the first time, add its entries to the dependency lock file, and mention in the `tofu init` output that it did so.
3. **Check that what you just installed is indeed what you intended to install**, since our installer uses a "trust on first use" model so until there's a dependency lock entry OpenTofu has no known-good selection to verify against.
4. Run `tofu plan`, `tofu apply`, or whatever other command you want that will now make use of the newly-installed plugin.

The important challenge with implementing _state storage_ as a plugin is that `tofu init` needs to fetch the state in order to learn which plugins it depends on, which means that it needs to configure the backend before the plugin installation step. Therefore as currently designed we have a chicken/egg problem where we need to have at least one provider installed before we can install the rest, _and_ we need to be able to actually execute that plugin directly in `tofu init` so that we can fetch the state using it.

---

I have a few potential ideas for how we could deal with this, but each has a user experience consequence so I think we'll want to consider these (and any other ideas others might add) quite carefully:

1. Introduce a special bootstrapping flow: if OpenTofu finds that it needs to install a plugin to fetch the state _and that plugin isn't yet recorded in the dependency lock file_ then `tofu init` would install just that one plugin, add it to the dependency lock file, and then immediately exit with a prompt to the user to verify what they just installed.

    If they are satisfied with what was installed then they can run `tofu init` a second time to complete the initialization. The second time it will already be installed and already be recorded in the dependency lock file, so we can execute it.

    Along with the awkwardness of requiring two steps during initial setup, this also introduces a bit of a hazard: today it's typically safe to just run `tofu init` at any time to resynchronize the working directory state with the configuration, and `tofu init` will either change something, succeed with a no-op if nothing needs to change, or will return an explicit error if the changes have been so significant as to require a user decision (such as whether to perform backend migration).

    Under the model I've described here, running `tofu init` a second time is treated as a signal that the user trusts the previously-installed backend plugin, which seems dangerous.

2. This option is really just a variation of the previous one, making it more explicit: when `tofu init` aborts early due to just having installed a not-previously-locked plugin, instead of immediately placing the plugin in the dependency lock file it would instead record this temporary state information in some other separate place and prompt the user to run `tofu init` again _with a specific extra command line argument_ to assert that they trust the plugin that was just installed. That second run would then add the plugin to the dependency lock file, and make immediate use of it to fetch the state.

3. Finally, instead of trying to extend the trust-on-first-use model to this new use-case we could conclude that state storage plugins deserve some greater scrutiny and must be pre-approved in some way by the OpenTofu project and signed with a key that is controlled by the OpenTofu project.

    This option seems like a non-starter to me given that the apparent goal is to open up an ecosystem for state storage plugins in a similar way as we have an ecosystem for provider plugins, but I've included it for completeness. This presumably means that state storage plugin developers would need to go through some "appstore-like" review process in order to publish new versions of their plugins, which feels culturally incongruent with everything else we're doing and time spent reviewing represents a significant opportunity cost competing with other work.

(Side-note: it's mildly amusing that the typical abbreviation of "trust on first use" is "TOFU". :upside_down_face: )

---

This also overlaps a little with [a comment I left on the other issue about workspaces in the HTTP backend](https://github.com/opentofu/opentofu/issues/317#issuecomment-2398262972), which was discussing the idea of using a predefined protocol that would be implemented in a remote server rather than a local plugin.

However, I think in practice we'd benefit from offering _both_, since they are suited to quite different use-cases: state storage plugins are good for adapting generic services (that were not designed specifically for OpenTofu) to use them for state storage, whereas a generalized protocol for the `remote` backend would be better suited to service providers that want to provide ergonomic OpenTofu state storage as a service, with the client-side implementation burden as small as possible and most of the logic implemented server-side.

(If anyone reading this has thoughts or concerns about _that_ idea then I'd encourage recording them over in https://github.com/opentofu/opentofu/issues/317 instead, so that we can keep this issue focused on the plugin-based approach. I just mentioned it here to draw attention to the overlap and differences between the use-cases.)

**Comment by apparentlymart** [comment]
Here's yet another thing that's on my mind as we think about reworking how backends work! :grimacing: 

It seems likely to me that introducing plugin-based state storage will mean at least _some_ syntax changes to allow for there being multiple registries and namespaces as with providers. (Perhaps backend plugins will be _part of_ provider plugins, or perhaps they will be something else, but either way I expect we'll want to have all of the same flexibility we currently have for provider plugin addressing and installation.)

If we're going to be making syntax changes _anyway_ then we could choose to address some other quirks of the current UX at the same time, and then keep the old syntax only as a deprecated alias for using one of the state storage implementations that was originally built in.

I want to say up front that I'm not intending this comment to mean that we _absolutely must_ block plugin-based state storage on designing a full revamp of the configuration. I'm just putting it on the table as an opportunity to potentially pay down some design debt in this area since it's rare to have an opportunity to redesign existing syntax.

---

The situation with backends today is quite strange, because the concepts and model presented to the user through the CLI and language don't actually really match how OpenTofu "thinks about" these concepts internally.

While there's nothing necessarily _wrong_ with having the UX be at a higher level of abstraction than the implementation -- indeed, that's often desirable -- in practice I've been intermittently answering community questions arising from this confusion for at least half a decade now :grimacing: so it seems like this particular abstraction isn't completely successful.

The main quirk today is that selecting a "backend" is really making two separate decisions:

1. Do I want to run everything locally, or do I want to integrate with a remote collaboration platform? ("TACOS")
2. If the answer to the previous question was "locally", then where do I want to store my state snapshots?

Internally in the code things have evolved in a zig-zaggy fashion of the years leading to lots of confusing terminology, but the most prominent way the code internally describes the above is: is the selected backend an "enhanced backend" (which takes over the whole workflow) or is it just a "state storage backend" (which uses the standard local workflow but provides an alternative storage location for state snapshots)? The documentation originally used similar terminology, but it didn't really seem to resonate with anyone except the person who originally chose it :upside_down_face: and so over time it was de-emphasized.

I think the main way this confusion shows up _in the UX_ is that there are two "backends" that both have special behavior that no other backend seems to have:

* `backend "remote"` was the original "integrate with a remote collaboration platform" backend, or "enhanced backend" if you prefer, which allows a remote service selected using a hostname to decide both whether your OpenTofu operations run locally or remotely _and_ where the resulting state snapshots should be stored.

    Despite this being presented as "just another backend", the explanation I've found that resonates best with people is that this effectively delegates the whole question of how Terraform behaves to a remote service, _instead of_ configuring your state storage backend locally.

* The `cloud` block is really just syntactic sugar for yet another backend called "cloud", which has much the same behavior as `backend "remote"` except that it was extended with newer behavior over time which `backend "remote"` doesn't have.

    So this one is _not_ presented as "just another backend", but internally it _is_ an "enhanced backend" much like `backend "remote"`.

Here's a summary of various different internal permutations created by writing different labels on the `backend` block:

| Config block       | Internal representation                                                                                                                                                                                                              |
|--------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| no block at all    | [`local.Local`](https://pkg.go.dev/github.com/opentofu/opentofu/internal/backend/local#Local) with `Backend` set to nil.                                                                                                             |
| `backend "local"`  | as above, but treated a little differently by `tofu init`'s init/migration logic.                                                                                                                                                    |
| `backend "s3"`     | [`local.Local`](https://pkg.go.dev/github.com/opentofu/opentofu/internal/backend/local#Local) with `Backend` set to [`s3.Backend`](https://pkg.go.dev/github.com/opentofu/opentofu@v1.8.4/internal/backend/remote-state/s3#Backend). |
| `backend "remote"` | [`remote.Remote`](https://pkg.go.dev/github.com/opentofu/opentofu@v1.8.4/internal/backend/remote#Remote), with the remote API deciding the state storage.                                                                            |
| `cloud`            | [`cloud.Cloud`](https://pkg.go.dev/github.com/opentofu/opentofu@v1.8.4/internal/cloud#Cloud), with the remote API deciding the state storage.                                                                                        |

Notice that `local.Local` has its own field `Backend` which has another nested `backend.Backend` in it. The nested backend is the "state storage backend", while the `local.Local` is the main backend that the `command` package thinks it's talking to.

If we're willing to invest in the backend and state storage configuration syntax while we're "in the neighborhood" preparing for plugins anyway, I think it'd be worth finding a way to eliminate the weird special cases around `backend "remote"` and `cloud` so that the user-facing mental model for backend configuration is a little more coherent.

----

So what might we do?

I don't have a fully-developed plan yet, but the main theme I'd like to address is more clearly separating the idea of "state storage", which would now always be dealt with using plugins, from "collaboration platforms" or "TACOS" which are integrated in a more streamline way and which effectively outsource all of the details about how OpenTofu workflow operations get run and where plan and state artifacts get stored _without_ the need to install any local plugins.

If I were starting from scratch today I expect I'd use the word "remote" to refer exclusively to the "collaboration platform" case and then use "state storage" to refer to the other case, giving two mutually-exclusive configuration blocks:

```hcl
# Remote workflow: no plugins required! Just specify the hostname of
# the platform you want to use and the platform-specific workspace selection
# settings.
#
# (NOTE: This doesn't _necessarily_ mean that the workflow operations will
# be run in a remote execution environment. It only gives the selected
# hostname the opportunity to decide whether to use local or remote
# operations, so it's valid to offer a collaboration service that only offers
# state/plan storage without also offering a remote execution environment.)

remote {
  # note: no label needed above! the hostname selects the implementation via network service discovery
  hostname     = "my-favorite-tacos.example.com"
  organization = "awesomecorp"

  workspaces {
    # The API at my-favorite-tacos.example.com gets to decide
    # what's valid in here and what it means, since different
    # services have different models for workspace selection:
    # labels vs. a fixed set of workspace names vs. workspace name
    # prefix or whatever else.
  }
}

# If you use a "remote" block then the "state_storage" block is forbidden, because
# the remote service gets to decide how state snapshots and saved plans are
# stored.
```

```hcl
# Local workflow: always runs all operations locally, but you can optionally
# choose a state storage plugin to use instead of the default filesystem-based
# implementation.

state_storage "registry.opentofu.org/happycloud/super-storage-service" {
  # The selected plugin has full control over what's valid in here and
  # what each setting means, just like today's backends.
}
```

Of course, we _aren't_ starting from scratch today, so perhaps we'd prefer something a little more familiar. In that case, the `remote` block I indicated above is essentially the same as the special `cloud` block, while the `state_storage` block is essentially the same as the `backend` block except that the label is now a fully-qualified plugin address instead of just a built-in name. If we want to go the path of reusing the existing terminology then I think that implies:

- Removing all of the remote-operations-related functionality from the `Backend` interface, making it exclusively focused on state storage, and then implementing that slimmed-down interface as a plugin client for state storage plugins.
- Deleting the current "remote" backend, since it's largely redundant with the cloud integration. (We'd probably choose to make it still work as an alias for a `cloud` block in practice, with a deprecation warning, for backward compatibility.)
- [Implementing a new vendor-agnostic API behind the `cloud` block](https://github.com/opentofu/opentofu/issues/317#issuecomment-2398262972), so that vendors can implement it without having to first reverse-engineer the initial vendor's proprietary API.

We could still decide that all of this is too much to take on as a prerequisite for backends-as-plugins! But let's consider the costs and benefits of doing a single overhaul of a bunch of UX design debt in a single step, so that there are fewer weird intermediate versions for users to learn about.

**Comment by ollevche** [comment]
Another note on the migration path and backward compatibility for backend implementations. 

Some of the backends may have additional assumptions around the `statemgr.Locker` interface. I've been recently reading through the `pg` backend implementation and it has a few interesting outlining behaviors (in comparison to other backends).

It doesn't use passed lock IDs at all: nor for `Lock`, neither for `Unlock` calls. Instead, it stores the cached version of the last acquired `LockInfo` in memory and then tries to release that exact lock regardless of what ID has been passed. This behavior makes sense since the locking mechanism is session-based (locks are released as soon as the connection drops).

In other words, `pg` backend is a stateful service on its own (not counting the db here) so it will have problems when failing independently of `tofu` binary.

I believe there could be more undocumented assumptions around the locking mechanism so we should carefully review each backend to ensure it can support "the new protocol" out of the box.

**Comment by vrmurthy** [comment]
Reviving this thread after a long gap. 
My team at Oracle is looking forward to this pluggable backend interface to be ready & implemented so that we can add support for our Oracle backend. This design makes it possible to add pluggable backends by any vendor and keeps it loosely coupled from the OpenTofu core.

**Comment by apparentlymart** [comment]
I notice we've not updated here for a while, so a note:

Folks have told us that the Terraform team has been sketching API designs for state storage as part of the Terraform plugin protocol. We don't want to ask provider developers to implement state storage twice to also support OpenTofu, so we're waiting a little to see what the final Terraform protocol looks like and, if it is something that OpenTofu can also use, to implement our own client for it so that vendors (like Oracle, above) could implement just one server implementation for both clients.

We are still hoping to design a more flexible and efficient model for state storage in the long run, at which point we probably will define our own plugin protocol tailored to that design, but we're hopeful that we'll be able to keep support for both so that providers using our own protocol will get the extra benefits but providers using the Terraform protocol would be able to offer state storage no worse than the current approach while being cross-compatible with Terraform.

All of this is still pretty early and subject to change, but I just wanted to keep you all updated on what's going on and where we think this is likely to go.

**Comment by cam72cam** [comment]
I've added a prototype for pluggable backends based on recent changes in the provider protocol.  It's a pathfinder and shows what areas of the code would need substantive changes.  https://github.com/opentofu/opentofu/pull/3476

**Comment by vmvarela** [comment]
Implemented OCI remote state backend support with compression, workspaces, and versioning (#3600). Awaiting final decision.

**Comment by cam72cam** [comment]
Now that https://github.com/hashicorp/terraform-plugin-go/pull/563 has been merged, we can start focusing on refining https://github.com/opentofu/opentofu/pull/3476 

As a note, we will wait for the provider protocol to be fully stabilized before we consider our implementation of this feature stable and ready for general use.

**Comment by mike-beliakov** [comment]
Me too comment

For now using the fork from @vmvarela for things like https://github.com/kvokka/opentaco-with-github-storage-example

**Comment by cam72cam** [comment]
Given that the terraform team is starting to make progress and [publicly documenting](https://developer.hashicorp.com/terraform/plugin/testing/acceptance-tests/state-store-mode) the format we will likely want to be compatible with, I'll bring this up with the maintainers team next triage to try to get this rolling.

Current progress can be seen in https://github.com/opentofu/opentofu/pull/3476
